### PR TITLE
Added the Buying limits to the grand exchange plugin.

### PR DESCRIPTION
**Pull request by psikoi**
- Added the display for the grand exchange buying limits to the side panel.
- Added a json file containing the names and respective limits (from wiki).
- Slightly modified the colours and phrasing on the GrandExchangeItemPanel.

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java [hunk 1]
```diff
@@ -79,15 +86,43 @@
...
+	private static Map<String, Integer> geLimitInfo;
+
+
 	@Provides
 	GrandExchangeConfig provideConfig(ConfigManager configManager)
 	{
 		return configManager.getConfig(GrandExchangeConfig.class);
 	}
 
+	private Map<String, Integer> loadGeLdimits()
```
Comment (SomeoneWithAnInternetConnection) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java]: Is this meant to be named `loadGeLimits`? It looks like an extra `d` slipped in there

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java [hunk 2]
```diff
@@ -79,15 +86,43 @@
...
+		}
+
+		return geLimitInfo.get(itemName);
+
+	}
+
 	@Override
 	protected void startUp() throws IOException
 	{
+		geLimitInfo = loadGeLdimits();
```
Comment (SomeoneWithAnInternetConnection) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java]: Setting up a static field in the instance's `startUp` method seems like not the best way to do this. Unfortunately I'm not entirely sure what the best alternative approach would be here. The only two that come to mind are: use a static initializer to set up the Map, or, scrap all the JSON stuff and add a `BuyLimit` enum, (with a method for lookups) instead

**Comment by psikoi** [comment]
Updating the PR:

- Fixed a typo in the loadGeLimits method

- Seperated the commit into three:
  - Logic (file loading, etc)
  - UI work to display the buying limits
  - UI work to update other g.e item displays

- Providing images of the final ui result:

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java [hunk 3]
```diff
@@ -79,15 +86,43 @@
 	@Inject
 	private GrandExchangeConfig config;
 
+	private static Map<String, Integer> geLimitInfo;
+
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java]: Too many new lines

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java [hunk 4]
```diff
@@ -79,15 +86,43 @@
...
 	}
 
+	private Map<String, Integer> loadGeLimits()
+	{
+		Gson gson = new Gson();
+		Type type = new TypeToken<Map<String, Integer>>()
+		{
+		}.getType();
+
+		InputStream limitsFile = GrandExchangePlugin.class.getResourceAsStream("/ge_limits.json");
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePlugin.java]: Would be nice to have API for this, but idk how useful would that be

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java [hunk 5]
```diff
@@ -114,11 +116,25 @@ public void mouseReleased(MouseEvent e)
...
-		haPriceLabel.setText(StackFormatter.formatNumber(haPrice.intValue()) + " alch");
-		haPriceLabel.setForeground(Color.orange);
-		rightPanel.add(haPriceLabel);
+		haPriceLabel.setText("Alch: " + StackFormatter.formatNumber(haPrice.intValue()));
+		haPriceLabel.setForeground(Color.gray);
+		detailsPanel.add(haPriceLabel);
+
+		// GE limit
+		double limit = GrandExchangePlugin.getLimit(name);
+		String limitIndicator = limit > -1 ? StackFormatter.formatNumber(limit) : "N/A";
```
[the remark this answers is not in the document]
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangeItemPanel.java]: I agree with using unknown instead of N/A

**Comment by psikoi** [comment]
Oh boy was I terrible.
I plan on coming back to this with a different approach, closing for now.

**Pull request closed without merging**